### PR TITLE
Add custom parameters to authorize and logout endpoint

### PR DESCRIPTION
**Pull request by eva-mueller-coremedia**
This PR
* provides a different approach to solve https://github.com/jenkinsci/oic-auth-plugin/issues/229 resp. https://github.com/jenkinsci/oic-auth-plugin/pull/241
* solves https://github.com/jenkinsci/oic-auth-plugin/issues/443

Additionally, some housekeeping (fixing typos, simply statements) has been done. 

The main change has been done in eb0d2ba1ed13aa1fd318fe3ad96cd876d49c7efb:
* Add new input fields for login resp. logout query parameters in the Jenkins management backend
* `maybeOpenIdLogoutEndpoint` has been refactored when it comes to combining all parameters
* Login query parameters are not allowed to overwrite keys as defined in `org.pac4j.oidc.config.OidcConfiguration`
* Logout query parameters are not allowed to overwrite keys as set in `maybeOpenIdLogoutEndpoint` like `id_token_hint`, `state`, `post_logout_redirect_uri`

### Testing done

This change has been tested by unit tests as well as local testing agains AWS Cognito.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by jtnord — COMMENTED** [review]
Hi @eva-mueller-coremedia thanks for this.

I have had a quick look and have a few comments around the UX side.

The extra parameters would IIUC not be needed in the majority of conformant IDPs, as such do you thing they should be hidden behind an [advanced section](https://reports.jenkins.io/core-taglib/jelly-taglib-ref.html#form:advanced)?

the use of a single string for multiple keyvalue pairs (and the encoding /splitting shenaningans) would seem to be error prone (and it lacks good warnings on invalid input).  (a value may need to encode `&` and then you need to escape it)
Would making the KeyValue a Describable and using a repeatable option using `repeatableProperty` or the like (which can then perform individual valuation of the key and the value).  NB not sur how this would look in CasC.

**Comment by eva-mueller-coremedia** [comment]
@jtnord Thanks for the feedback, will try to follow a key value pair pattern.

If you don't mind, I would like to remove the UI at all - only provide it to be set via CasC resp. Java/Groovy code. Any objections?

**Comment by jtnord** [comment]
> @jtnord Thanks for the feedback, will try to follow a key value pair pattern.
> 
> If you don't mind, I would like to remove the UI at all - only provide it to be set via CasC resp. Java/Groovy code. Any objections?

it's generally considered bad to have something exposed in the UI and not CasC and vice versa.  Not exposing in the UI can lead to it getting overridden if someone saves the Jenkins form in the UI (which I say you should never do as if you manage by code you should have a read only system UI, but it is useful on test systems where you want to make changes in the UI and then export the config for applying to a different server).

**Comment by eva-mueller-coremedia** [comment]
> > @jtnord Thanks for the feedback, will try to follow a key value pair pattern.
> > If you don't mind, I would like to remove the UI at all - only provide it to be set via CasC resp. Java/Groovy code. Any objections?
> 
> it's generally considered bad to have something exposed in the UI and not CasC and vice versa. Not exposing in the UI can lead to it getting overridden if someone saves the Jenkins form in the UI (which I say you should never do as if you manage by code you should have a read only system UI, but it is useful on test systems where you want to make changes in the UI and then export the config for applying to a different server).

Understood. 👍  I will have a look to implement it the way you described initially. Thanks for the explanation!

File: src/main/java/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration.java [hunk 1]
```diff
@@ -219,12 +219,12 @@ public FormValidation doCheckWellKnownOpenIDConfigurationUrl(
         }
 
         @POST
-        public FormValidation doCheckOverrideScopes(@QueryParameter String overrideScopes) {
+        public FormValidation doCheckScopesOverride(@QueryParameter String scopesOverride) {
```
Comment (eva-mueller-coremedia) [inline src/main/java/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration.java:222]: @jtnord The form validation for scopes did not work since the jelly uses [`scopesOverride`](https://github.com/jenkinsci/oic-auth-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration/config.jelly#L10) as field name. 

Fixed this with this PR or should I open a dedicated PR?

**Comment by eva-mueller-coremedia** [comment]
@jtnord Thanks for the review. Will take care after my 1 week vacation.

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 2]
```diff
@@ -505,7 +533,7 @@ ProxyAwareResourceRetriever getResourceRetriever() {
         return proxyAwareResourceRetriever;
     }
 
-    private OidcConfiguration buildOidcConfiguration() {
+    private OidcConfiguration buildOidcConfiguration(boolean addCustomLoginParams) {
```
[the remark this answers is not in the document]
Comment (eva-mueller-coremedia) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:536]: I thought of this as kind of "optimization" but looking at this some days later and with your review comment, I will dissolve this.

File: src/main/java/org/jenkinsci/plugins/oic/OicQueryParameterConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    public String getQueryParamNameDecoded() {
+        return paramName != null
+                ? URLEncoder.encode(paramName, StandardCharsets.UTF_8).trim()
+                : null;
+    }
+
+    public String getQueryParamValueDecoded() {
+        return paramValue != null
+                ? URLEncoder.encode(paramValue, StandardCharsets.UTF_8).trim()
```
[the remark this answers is not in the document]
Comment (eva-mueller-coremedia) [reply src/main/java/org/jenkinsci/plugins/oic/OicQueryParameterConfiguration.java:63]: Oh, thanks for pointing to this. Will fix the method name

File: src/main/java/org/jenkinsci/plugins/oic/OicQueryParameterConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    public String getQueryParamValue() {
+        return paramValue;
+    }
+
+    public String getQueryParamNameDecoded() {
+        return paramName != null
+                ? URLEncoder.encode(paramName, StandardCharsets.UTF_8).trim()
+                : null;
+    }
```
[the remark this answers is not in the document]
Comment (eva-mueller-coremedia) [reply src/main/java/org/jenkinsci/plugins/oic/OicQueryParameterConfiguration.java:59]: Oh, thanks for pointing to this. Will fix the method name

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 5]
```diff
@@ -357,6 +364,9 @@ protected Object readResolve() throws ObjectStreamException {
         // ensure escapeHatchSecret is encrypted
         this.setEscapeHatchSecret(this.escapeHatchSecret);
 
+        this.setLoginQueryParamNameValuePairs(this.loginQueryParamNameValuePairs);
+        this.setLogoutQueryParamNameValuePairs(this.logoutQueryParamNameValuePairs);
```
[the remark this answers is not in the document]
Comment (eva-mueller-coremedia) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:368]: Yes, seem to be leftover. Will remove the calls

File: src/main/java/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration.java (hunk 1 above)
Comment (eva-mueller-coremedia) [reply src/main/java/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration.java:222]: I will open a dedicated PR for this.

**Comment by eva-mueller-coremedia** [comment]
@jtnord How to re-trigger the failed Jenkins run - there has been a `java.net.SocketTimeoutException: Read timed out` when downloading artifacts from azure...

**Comment by jtnord** [comment]
> @jtnord How to re-trigger the failed Jenkins run - there has been a `java.net.SocketTimeoutException: Read timed out` when downloading artifacts from azure...

For non maintainers, adding a new commit.

**Comment by eva-mueller-coremedia** [comment]
@jtnord The PR is now ready for a re-review.

**Comment by eva-mueller-coremedia** [comment]
Sorry for nudging again - @jtnord @michael-doubez would be great if you could have a look.

**Review by jtnord — COMMENTED** [review]
The code looks reasonable but some things look a little weird and I am wondering if thebUX can be improved at all.
I wanted to try and get some time to see if this can be improved and submit a PR to the PR as doing so after merging would possibly require even more data migration that we already have.realistically this is not going to happen this week.

**Comment by eva-mueller-coremedia** [comment]
> realistically this is not going to happen this week.

Thanks for this estimate. Would be great to see this feature in the (near) future since AWS Cognito requires some additional query parameters.

It would be great if you could keep me/subscribers posted about the state of this PR from time to time. Thanks in advance.

**Comment by jtnord** [comment]
@eva-mueller-coremedia I filed 
https://github.com/eva-mueller-coremedia/oic-auth-plugin/pull/1 against your PR.

Moves the parameters to their own types and added validation for them so it's validated in the UI.  Also ensures the config can never be invalid.
Tidies up the jelly slightly.

File: src/main/java/org/jenkinsci/plugins/oic/LoginQueryParameter.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    @DataBoundConstructor
+    public LoginQueryParameter(String key, String value) throws FormException {
+        super(key, value);
+    }
+
+    @Extension
+    public static class DescriptorImpl extends AbstractKeyValueDescribable.DescriptorImpl<LoginQueryParameter> {
+
+        @Override
+        public FormValidation doCheckKey(@QueryParameter String key) {
```
[the remark this answers is not in the document]
Comment (jtnord) [reply src/main/java/org/jenkinsci/plugins/oic/LoginQueryParameter.java:21]: There is nothing sensitive here.
Whilst we could add a permission check we are only doing string comparisons

[the remark this answers is not in the document]
Comment (eva-mueller-coremedia) [reply src/main/java/org/jenkinsci/plugins/oic/LoginQueryParameter.java:21]: @jtnord Unfortunately, I do not know how to resolve all the problems found by the github-advanced-security. Can you help me here or step in? That would be great.

Also there is already one dismissed problem. Dunno who or how this got dismissed.

**Comment by eva-mueller-coremedia** [comment]
> @eva-mueller-coremedia I filed [eva-mueller-coremedia#1](https://github.com/eva-mueller-coremedia/oic-auth-plugin/pull/1) against your PR.
> 
> Moves the parameters to their own types and added validation for them so it's validated in the UI. Also ensures the config can never be invalid. Tidies up the jelly slightly.

Thank you for the PR. I removed the commented out test.

File: src/main/java/org/jenkinsci/plugins/oic/LoginQueryParameter.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jtnord) [reply src/main/java/org/jenkinsci/plugins/oic/LoginQueryParameter.java:21]: I can take a look.
I dismissed one as a false positive. (`key` was not sensitive in that context)

File: src/main/java/org/jenkinsci/plugins/oic/AbstractKeyValueDescribable.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        FormValidation keyValidation = getDescriptor().doCheckKey(key);
+        if (keyValidation.kind == Kind.ERROR) {
+            throw new Descriptor.FormException(keyValidation.getMessage(), "key");
+        }
+        FormValidation valueValidation = getDescriptor().doCheckValue(value);
+        if (valueValidation.kind == Kind.ERROR) {
+            throw new Descriptor.FormException(valueValidation.getMessage(), "value");
+        }
+        this.key = key.trim();
+        this.value = value == null ? "" : value.trim();
```
Comment (eva-mueller-coremedia) [inline src/main/java/org/jenkinsci/plugins/oic/AbstractKeyValueDescribable.java:52]: Hi @jtnord - I added some more test cases and encountered for LogoutQueryParameters a NPE when the value is null.

Is this solution OK for you?

**Comment by eva-mueller-coremedia** [comment]
@jtnord @michael-doubez @basil - May I politely ask if there is a change that this PR get's reviewed resp. merged soon or are you busy with other things right now? There for nudging again...
